File: Lab5_middleware/SmartHome/src/client/devices/lightbulb.py
from devices import device_handler_loop


def turn_on(lightbulb):
    lightbulb.turnOn()


def turn_off(lightbulb):
    lightbulb.turnOff()


def get_color(lightbulb):
    print("Current color: {}".format(lightbulb.getColor()))


def get_lightbulb_state(lightbulb):
    print("Current state: {}".format(lightbulb.getLightBulbState()))


def get_device_info(lightbulb):
    print("Device info: {}".format(lightbulb.getDeviceInfo()))


def set_color(lightbulb):
    color = input("new color: ")
    lightbulb.setColor(color)
    get_color(lightbulb)


operations = {
    "info": get_device_info,
    "turnOn": turn_on,
    "turnOff": turn_off,
    "getColor": get_color,
    "setColor": set_color,
    "getState": get_lightbulb_state,
    "exit": None
}


def lightbulb_handler(lightbulb, device_name):
    device_handler_loop.run_loop(operations, lightbulb, device_name)

File: Lab5_middleware/SmartHome/server_java/src/server/devices/LightBulb.java
package devices;

import SmartHome.*;
import com.zeroc.Ice.Current;

public class LightBulb implements SmartHome.LightBulb {
    protected final DeviceInfo deviceInfo;
    protected final LightBulbState lightbulbState;

    public LightBulb(DeviceInfo info) {
        this.deviceInfo = info;
        this.lightbulbState = new LightBulbState();
    }

    @Override
    public DeviceInfo getDeviceInfo(Current current) {
        return deviceInfo;
    }

    @Override
    public LightBulbState getLightBulbState(Current current) {
        return lightbulbState;
    }

    @Override
    public Color getColor(Current current) throws DeviceTurnedOffError {
        if (!lightbulbState.tunredOn) {
            throw new DeviceTurnedOffError("LightBulb " + deviceInfo.name + " is turned off");
        }
        return lightbulbState.color;
    }

    @Override
    public void setColor(String newColor, Current current) throws DeviceTurnedOffError, InvalidColorError {
        if (!lightbulbState.tunredOn) {
            throw new DeviceTurnedOffError("LightBulb " + deviceInfo.name + " is turned off");
        }
        try {
            lightbulbState.color = Color.valueOf(newColor.toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new InvalidColorError("Color " + newColor + " is not a valid color name");
        }
    }

    @Override
    public void turnOn(Current current) {
        lightbulbState.tunredOn = true;
    }

    @Override
    public void turnOff(Current current) {
        lightbulbState.tunredOn = false;
    }
}
